test/common/websocket/codec_test.cc

```diff
@@ -0,0 +1,824 @@
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(false, frames.value()[0].masking_key_.has_value());
+  EXPECT_EQ(1, frames.value().size());
+  EXPECT_EQ(0, frames.value()[0].payload_length_);
+}
+
+TEST(WebSocketCodecTest, encodeFrameHeader) {
+  Encoder encoder;
+  std::vector<uint8_t> buffer;
+  std::vector<uint8_t> expected_header;
+
+  buffer.clear();
+  buffer = encoder.encodeFrameHeader({true, 1, absl::nullopt, 5, nullptr});
+  EXPECT_EQ(buffer.size(), 2);
+  expected_header = {0x81, 0x05};
+  EXPECT_EQ(buffer, expected_header);
+
+  buffer.clear();
+  buffer = encoder.encodeFrameHeader({true, 1, 0x37fa213d, 5, nullptr});
+  EXPECT_EQ(buffer.size(), 6);
+  expected_header = {0x81, 0x85, 0x37, 0xfa, 0x21, 0x3d};
+  EXPECT_EQ(buffer, expected_header);
+
+  buffer.clear();
+  buffer = encoder.encodeFrameHeader({false, 0, 0x3c332a16, 5, nullptr});
+  EXPECT_EQ(buffer.size(), 6);
+  expected_header = {0x00, 0x85, 0x3c, 0x33, 0x2a, 0x16};
+  EXPECT_EQ(buffer, expected_header);
+
+  buffer.clear();
+  buffer = encoder.encodeFrameHeader({true, 2, absl::nullopt, 256, nullptr});
+  EXPECT_EQ(buffer.size(), 4);
+  expected_header = {0x82, 0x7e, 0x01, 0x00};
+  EXPECT_EQ(buffer, expected_header);
+
+  buffer.clear();
+  buffer = encoder.encodeFrameHeader({true, 2, 0x37fa213d, 256, nullptr});
+  EXPECT_EQ(buffer.size(), 8);
+  expected_header = {0x82, 0xfe, 0x01, 0x00, 0x37, 0xfa, 0x21, 0x3d};
+  EXPECT_EQ(buffer, expected_header);
+
+  buffer.clear();
+  buffer = encoder.encodeFrameHeader({true, 2, absl::nullopt, 77777, nullptr});
+  EXPECT_EQ(buffer.size(), 10);
+  expected_header = {0x82, 0x7f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x2f, 0xd1};
+  EXPECT_EQ(buffer, expected_header);
+
+  buffer.clear();
+  buffer = encoder.encodeFrameHeader({true, 2, 0x37fa213d, 77777, nullptr});
+  EXPECT_EQ(buffer.size(), 14);
+  expected_header = {0x82, 0xff, 0x00, 0x00, 0x00, 0x00, 0x00,
+                     0x01, 0x2f, 0xd1, 0x37, 0xfa, 0x21, 0x3d};
+  EXPECT_EQ(buffer, expected_header);
+}
+
+TEST(WebSocketCodecTest, decodeClosingFrame) {
+  Encoder encoder;
+  Buffer::OwnedImpl buffer;
+  std::vector<uint8_t> header;
+
+  header = encoder.encodeFrameHeader({false, 8, absl::nullopt, 0, nullptr});
+  buffer.add(header.data(), header.size());
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(1, frames.value().size());
+  EXPECT_EQ(0, frames.value()[0].payload_length_);
+  EXPECT_EQ(8, frames.value()[0].opcode_);
+  EXPECT_FALSE(frames.value()[0].final_fragment_);
+}
+
+TEST(WebSocketCodecTest, decodeInvalidFrame) {
+  Encoder encoder;
+  Buffer::OwnedImpl buffer;
+  std::vector<uint8_t> header;
+  // Invalid opcode
+  header = encoder.encodeFrameHeader({false, 0x0b, absl::nullopt, 3, nullptr});
+  buffer.add(header.data(), header.size());
+  buffer.add("Hey");
```

Comment:
Since this is a Decoder test, I don't think we really need the Encoder, do we? Can we instead just provide the literal frame input like we do for the other tests?

At the same time, this test makes me wonder if the Encoder should really really be serialized an invalid frame? It makes me think that it should possibly error out of the opcode is invalid. What do you think?

Reply:
yes it should validate it. opcode validation added to both encoder and decoder and error log also added.